sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala

```diff
@@ -91,10 +111,19 @@ case class Percentile(
     case _ => DoubleType
   }
 
-  override def inputTypes: Seq[AbstractDataType] = percentageExpression.dataType match {
-    case _: ArrayType => Seq(NumericType, ArrayType(DoubleType))
-    case _ => Seq(NumericType, DoubleType)
-  }
+  override def inputTypes: Seq[AbstractDataType] =
+    if (frequency == unit) {
+      percentageExpression.dataType match {
+        case _: ArrayType => Seq(NumericType, ArrayType(DoubleType))
+        case _            => Seq(NumericType, DoubleType)
+      }
+    } else {
+      percentageExpression.dataType match {
```

Comment:
It is not needed to make a difference between freq = 1 and the freq > 1 cases. Just use the more complete case.